cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```
@@ -0,0 +1,134 @@
+ * the License.
+ */
+
+package io.cdap.cdap.sourcecontrol.operationrunner;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.inject.Inject;
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import io.cdap.cdap.sourcecontrol.CommitMeta;
+import io.cdap.cdap.sourcecontrol.NoChangesToPushException;
+import io.cdap.cdap.sourcecontrol.RepositoryManager;
+import io.cdap.cdap.sourcecontrol.RepositoryManagerFactory;
+import io.cdap.cdap.sourcecontrol.UnexpectedRepositoryChangesException;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+
+/**
+ * In-Memory implementation for {@link SourceControlOperationRunner}.
+ * Runs all git operation inside calling service.
+ */
+public class InMemorySourceControlOperationRunner implements SourceControlOperationRunner {
+  private final RepositoryManagerFactory repoManagerFactory;
+  private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+  private static final Logger LOG = LoggerFactory.getLogger(InMemorySourceControlOperationRunner.class);
+
+  @Inject
+  public InMemorySourceControlOperationRunner(RepositoryManagerFactory repoManagerFactory) {
+    this.repoManagerFactory = repoManagerFactory;
+  }
+
+  @Override
+  public PushAppResponse push(NamespaceId namespace, RepositoryConfig repoConfig,
+                              ApplicationDetail appToPush, CommitMeta commitDetails)
+    throws Exception {
+    try (RepositoryManager repositoryManager = repoManagerFactory.create(namespace, repoConfig)) {
+      try {
+        repositoryManager.cloneRemote();
+      } catch (Exception e) {
+        throw new PushFailureException("Failed to clone remote repository", e);
```

Any time we catch an exception and re-throw it, we should be checking if that's really the right thing to do. For example, this message is not going to be useful if displayed to an end user. Can we instead just let the exception propagate up? What types of exceptions can cloneRemote() throw?

---

It throws `IOException, AuthenticationConfigException, GitAPIException`. We should catch `GitAPIException` at least because it's a jgit dependency and should not be propagated to service layer